packages/loader/container-loader/src/protocol.ts

```diff
@@ -0,0 +1,101 @@
+    IProtocolHandler as IBaseProtocolHandler,
+    IQuorumSnapshot,
+    ProtocolOpHandler,
+} from "@fluidframework/protocol-base";
+import {
+    IDocumentAttributes,
+    IProcessMessageResult,
+    ISequencedDocumentMessage,
+    ISignalClient,
+    ISignalMessage,
+    MessageType,
+} from "@fluidframework/protocol-definitions";
+
+export interface IProtocolDetails {
+    /**
+     * Function to be used for creating a protocol handler.
+     */
+    protocolHandlerBuilder: ProtocolHandlerBuilder;
+}
+
+export type ProtocolHandlerBuilder = (
+    attributes: IDocumentAttributes,
+    snapshot: IQuorumSnapshot,
+    sendProposal: (key: string, value: any) => number,
+    initialClients: ISignalClient[],
+) => IProtocolHandler;
+
+export interface IProtocolHandler extends IBaseProtocolHandler {
+    readonly audience: IAudienceWriter;
+    // To be removed after the server package dependency is upgraded. ADO:1026
+    processSignal(message: ISignalMessage);
+}
+
+export class ProtocolHandler extends ProtocolOpHandler implements IProtocolHandler {
+    constructor(
+        attributes: IDocumentAttributes,
+        quorumSnapshot: IQuorumSnapshot,
+        sendProposal: (key: string, value: any) => number,
+        initialClients: ISignalClient[],
+        readonly audience: IAudienceWriter,
+    ) {
+        super(
+            attributes.minimumSequenceNumber,
+            attributes.sequenceNumber,
+            attributes.term,
+            quorumSnapshot.members,
+            quorumSnapshot.proposals,
+            quorumSnapshot.values,
+            sendProposal,
+        );
+
+        for (const initialClient of initialClients) {
+            this.audience.addMember(initialClient.clientId, initialClient.client);
+        }
+    }
+
+    public processMessage(message: ISequencedDocumentMessage, local: boolean): IProcessMessageResult {
+        const client: ILocalSequencedClient | undefined = this.quorum.getMember(message.clientId);
+
+        // Check and report if we're getting messages from a clientId that we previously
+        // flagged as shouldHaveLeft, or from a client that's not in the quorum but should be
+        if (message.clientId != null) {
+            if (client === undefined && message.type !== MessageType.ClientJoin) {
+                // pre-0.58 error message: messageClientIdMissingFromQuorum
+                throw new Error("Remote message's clientId is missing from the quorum");
+            }
+
+            if (client?.shouldHaveLeft === true && message.type !== MessageType.NoOp) {
+                // pre-0.58 error message: messageClientIdShouldHaveLeft
+                throw new Error("Remote message's clientId already should have left");
+            }
+        }
+
+        return super.processMessage(message, local);
+    }
+
+    public processSignal(message: ISignalMessage) {
+        const innerContent = message.content as { content: any; type: string; };
+        switch (innerContent.type) {
```

Comment:
This code seems like should belong to Audience class (i.e. processSignal() should call this.audience.processSignal()

Reply:
I'll look into that, thanks!

Reply:
I'm not sure, actually.

The responsibility of the Audience class is to handle membership. If the membership is driven by signals/messages or other entities, I think that should be in the protocol handler and the Audience shouldn't be responsible for unpacking messages or even know about messages.